DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/shannonization/src/shannonization.cpp

```diff
@@ -0,0 +1,335 @@
+
+  // setup the output buffer
+  buffer<int, 1> n_buf(&n, 1);
+
+  // submit the kernel that produces table A
+  q.submit([&](handler& h) {
+    accessor a_accessor { a_buf, h, read_only };
+    h.single_task<ProducerA<Version>>([=]() [[intel::kernel_args_restrict]] {
+      for (int i = 0; i < a_size; i++) {
+        ProduceAPipe::write(a_accessor[i]);
+      }
+    });
+  });
+
+  // submit the kernel that produces table B
+  q.submit([&](handler& h) {
+    accessor b_accessor { b_buf, h, read_only };
+    h.single_task<ProducerB<Version>>([=]() [[intel::kernel_args_restrict]] {
+      for (int i = 0; i < b_size; i++) {
+        ProduceBPipe::write(b_accessor[i]);
+      }
+    });
+  });
+
+  // submit the kernel that performs the intersection
+  event e = q.submit([&](handler& h) {
+    // output accessor
+    accessor n_accessor { n_buf, h, write_only, noinit };
+
+    h.single_task<Worker<Version>>([=]() [[intel::kernel_args_restrict]] {
+      // The 'Version' template parameter will choose between the different 
+      // versions of the kernel defined in IntersectionKernel.hpp.
+      // The operator() of the IntersectionKernel object will return the
+      // size of the intersection of A and B.
+      IntersectionKernel<Version, II, ProduceAPipe, ProduceBPipe> K;
+      n_accessor[0] = K(a_size, b_size);
+    });
+  });
+
+  return e;
+}
+
+//
+// This function performs the intersection by submitting the 
+// different kernels. This method also validates the output
+// of the kernels and prints performance information
+//
+template <int Version, int II>
+bool Intersection(queue& q, std::vector<unsigned int>& a, 
+                  std::vector<unsigned int>& b, int golden_n) {
+  // For emulation, just do a single iteration.
+  // For hardware, perform multiple iterations for a more
+  // accurate throughput measurement
+#if defined(FPGA_EMULATOR)
+  int iterations = 1;
+#else
+  int iterations = 5;
+#endif
+
+  std::cout << "Running " << iterations 
+            << ((iterations == 1) ? " iteration" : " iterations")
+            << " of kernel " << Version
+            << " with |A|=" << a.size() 
+            << " and |B|=" << b.size() << "\n";
+  
+  bool success = true;
+  std::vector<double> kernel_latency(iterations);
+
+  // perform multiple iterations of the kernel to get a more accurate
+  // throughput measurement
+  for (size_t i = 0; i < iterations && success; i++) {
+    // run kernel
+    int n = 0;
+    event e = SubmitKernels<Version,II>(q, a, b, n);
+
+    // check output
+    if (golden_n != n) {
+      success = false;
+      std::cerr << "ERROR: Kernel version " << Version << " output is incorrect"
+                << " (Expected=" << golden_n << ", Result=" << n << ")\n";
+    }
+
+    // get profiling info
+    auto start = e.get_profiling_info<info::event_profiling::command_start>();
+    auto end = e.get_profiling_info<info::event_profiling::command_end>();
+    kernel_latency[i] = (end - start) / 1000000000.0f;
```

Comment:
```suggestion
    kernel_latency[i] = (end - start) / 1e9;
```
I R=removed also the `f` since the result is a... `double`!